packages/remirror__extension-search/src/search-extension.ts

```diff
@@ -306,86 +116,103 @@ export class SearchExtension extends PlainExtension<SearchOptions> {
       }
 
       tr.insertText(replacement, result.from, result.to);
-      return this.searchNext()(props);
+      return this.updateView(tr, dispatch);
     };
   }
 
-  private rebaseNextResult({ replacement, index, lastOffset = 0 }: RebaseNextResultProps) {
-    const nextIndex = index + 1;
-
-    if (!this._results[nextIndex]) {
-      return;
-    }
+  @command()
+  replaceAllSearchResults({ replacement }: ReplaceAllOptions): CommandFunction {
+    return ({ tr, dispatch }) => {
+      const ranges = this.gatherSearchResults(tr.doc);
 
-    const current = assertGet(this._results, index);
-    const offset = current.to - current.from - replacement.length + lastOffset;
-    const next = assertGet(this._results, nextIndex);
+      for (let i = ranges.length - 1; i >= 0; i--) {
+        const { from, to } = ranges[i];
+        tr.insertText(replacement, from, to);
+      }
 
-    this._results[nextIndex] = {
-      to: next.to - offset,
-      from: next.from - offset,
+      this._searchTerm = '';
+      this._ranges = [];
+      this._activeIndex = undefined;
+      return this.updateView(tr, dispatch);
     };
-
-    return offset;
   }
 
-  private replaceAll(replacement: string): CommandFunction {
-    return (props) => {
-      const { tr, dispatch } = props;
-      let offset: number | undefined;
-
-      if (isEmptyArray(this._results)) {
-        return false;
-      }
+  /**
+   * This plugin is responsible for do the searching and updating the
+   * decorations.
+   */
+  createPlugin(): CreateExtensionPlugin {
+    return {
+      state: {
+        init() {
+          return DecorationSet.empty;
+        },
+        apply: (tr, old) => {
+          if (this._updating || (tr.docChanged && this.options.alwaysSearch)) {
+            const doc = tr.doc;
+            this._ranges = this.gatherSearchResults(doc);
```

Comment:
Is searching the whole doc on each transaction expensive?

Could consider gathering the changed nodes and searching those?

Reply:
> Is searching the whole doc on each transaction expensive?

If the document is big, yes. 

> Could consider gathering the changed nodes and searching those?

That would be great. However I'd like to firstly finish the features (including the React search component) and then improve the performance. 

--- 

After a further research, I find that the browser's native search function is more similar to setting `alwaysSearch` as `false`. 

So I decide to set the default value of `alwaysSearch` as `false`. This will improve the default performance and give a not bad result.  

**Browser native search:**

https://user-images.githubusercontent.com/24715727/197514507-5bcb1289-1575-4f1b-89a6-a8281a828820.mp4


**`SearchExtension({alwaysSearch: false})`**:

https://user-images.githubusercontent.com/24715727/197514618-f58523b7-037e-4e60-9bf1-61f40a18ffd7.mp4


**`SearchExtension({alwaysSearch: true})`**:

https://user-images.githubusercontent.com/24715727/197514632-8ce31ffe-3498-4c36-9dab-40ba72281284.mp4